contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1024 @@
+        role: "SCHEDULE_CREATOR_ROLE"
+      });
+    }
+    if (_scheduleIdToScheduleStruct[projectId].exists) {
+      revert ScheduleExists({scheduleTokenId: projectId});
+    }
+    Schedule storage schedule = _scheduleIdToScheduleStruct[projectId];
+    _allScheduleIds.add(projectId);
+    RestrictionDuration
+      memory restrictionDuration = getRestrictionDurationForMethodologyAndVersion(
+        scheduleData.methodology,
+        scheduleData.methodologyVersion
+      );
+    if (!restrictionDuration.wasSet) {
+      revert RestrictionDurationNotSet({projectId: projectId});
+    }
+    _addressToScheduleIdSet[recipient].add(projectId);
+    schedule.exists = true;
+    schedule.startTime = scheduleData.startTime;
+    schedule.endTime = scheduleData.startTime + restrictionDuration.duration;
+    schedule.releasedAmountFloor = 0;
+    emit ScheduleCreated(projectId);
+  }
+
+  /**
+   * @notice Revokes unreleased tokens from a schedule.
+   * This is an *admin* operation callable only by addresses having TOKEN_REVOKER_ROLE
+   * (enforced in `batchRevokeUnreleasedTokenAmounts`)
+   *
+   * @dev Only the maximum revocable number of tokens (unreleased tokens) can be revoked.
+   * Once the tokens have been revoked, the current released amount can never fall below
+   * its current level, even if the linear release schedule of the new amount would cause
+   * the released amount to be lowered at the current timestamp (a floor is established).
+   *
+   * Unlike in the `withdrawFromSchedule` function, here we burn `RestrictedNORI`
+   * from the schedule owner but send that `BridgedPolygonNORI` back to Nori's
+   * treasury or an address of Nori's choosing (the *to* address).
+   * The *claimedAmount* is not changed because this is not a claim operation.
+   */
+  function _revokeUnreleasedTokens(
+    address to,
+    uint256 removalId,
+    uint256 amount
+  ) internal {
+    uint256 scheduleId = _removal.getProjectIdForRemoval(removalId);
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    if (!schedule.exists) {
+      revert NonexistentSchedule({scheduleId: scheduleId});
+    }
+    uint256 quantityToRevoke;
+    // amount of zero indicates revocation of all remaining tokens.
+    if (amount > 0) {
+      if (!(amount <= revocableQuantityForSchedule(scheduleId))) {
+        revert InsufficientUnreleasedTokens({scheduleId: scheduleId});
+      }
+      quantityToRevoke = amount;
+    } else {
+      quantityToRevoke = revocableQuantityForSchedule(scheduleId);
+    }
+
+    // burn correct proportion from each token holder
+    EnumerableSetUpgradeable.AddressSet storage tokenHolders = schedule
+      .tokenHolders;
+    uint256[] memory quantitiesToBurnForHolders = new uint256[](
+      tokenHolders.length()
+    );
+
+    uint256 cumulativeQuantityToBurn = 0;
+    for (uint256 i = 0; i < (tokenHolders.length() - 1); i++) {
+      uint256 quantityToBurnForHolder = _quantityToRevokePerTokenHolder(
```

Comment:
Maybe would simplify this also to call out to get the holder's percentage rather than outsourcing the actual revocable amount calculation?

Reply:
precision problems